[oneDAL] Add common alg descriptor spec (#145)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/programming_model/algorithm_anatomy.rst

.. highlight:: cpp
.. default-domain:: cpp

=================
Algorithm anatomy
=================

.. _descriptors:

-----------
Descriptors
-----------

**Descriptor** is an object that represents an algorithm including all its
properties and dependencies on other algorithms. Descriptor serves as:

- Dispatching mechanism for the `operations`_. Depending on descriptor
  type, operation decides what algorithm implementation should be executed.

- Aggregator of parameters. Descriptor provides an interface for setting up both
  compile-time and run-time parameters.

- Object that carries state of the algorithm. In the general case, descriptor is
  stateful object whose state changes after applying an operation.

Each oneDAL algorithm has its own dedicated namespace, where corresponding

> descriptor is defined. Descriptor, in its turn, defines a set of template

Reviewer: bullet list works fine

parameters and **properties** --- run-time parameters can be accessed by means

> of the corresponding getter and setter methods. `The following code sample

Author: I'm going to add them to glossary once it's merged to master

   public:
      /* Getter & Setter for the property called `%PROPERTY_NAME%` */
      descriptor& set_%PROPERTY_NAME%(%PROPERTY_TYPE% value);
      %PROPERTY_TYPE% get_%PROPERTY_NAME%() const;

      /* more properties */
   };

   } // namespace onedal::%ALGORITHM%

Descriptor may define any number of properties. Most of the properties are
preset with default values. Some of the properties are initialized by passing
required parameters to the constructor.

Properties are defined by means of getter and setter method. The getter methods
returns value of that type, while the setter accepts it as a parameter. The
underlying class field that stores the property's value is never exposed in the
descriptor interface. In addition, the setter returns a reference to the
descriptor object to allow chaining calls as shown in the example below.

.. code-block:: cpp
   :caption: Example of chaining calls to set some property values to the descriptor

   auto desc = descriptor{}
      .set_property_name_1(value_1)
      .set_property_name_2(value_2)
      .set_property_name_3(value_3);

Descriptor is allowed to have any number of template parameters, but must
support at least two:

> - ``Float`` is a floating-point type that the algorithm uses for computations.

Reviewer 2: This paragraph seems to deserve a separate dedicated section that would be discuss data types, etc and this default_float_t living in the library level namespace

  Algorithms are required to support only ``float`` and ``double`` types. The
  other floating-point types, e.g., ``float16`` or ``bfloat16``, are optional
  and up to specific oneDAL implementation. This parameter is defined first and
  has the :expr:`onedal::default_float_t` default value.

> - ``Method`` is a tag-type that specifies an implementation of algorithm.

Reviewer 2: The same about this method info - dedicated section, and the descriptor would then refer to them and only specify the order of the template parameters

  Computational scheme of the algorithm may be implemented in different ways, so
  this parameter specifies the method to use while some operation is performed.

  The supported computational methods are declared within the ``method``

>   namespace nested to the algorithm's namespace. Algorithm must support at least

Reviewer 2: must->shall

Reviewer: shall is more like will, should is better

`descriptors`_ and `operations`_.

.. _descriptors:

-----------
Descriptors
-----------

**Descriptor** is an object that represents an algorithm including all its
meta-parameters, dependencies on other algorithms, floating-point types, and
computational methods. Descriptor serves as:

- Dispatching mechanism for the `operations`_. Depending on descriptor
  type, operation decides what algorithm implementation should be executed.

- Aggregator of meta-parameters. Descriptor provides an interface for setting up
  meta-parameters in compile-time or run-time.

- Object that carries state of the algorithm. In the general case, descriptor is
  stateful object whose state changes after applying an operation.

Each oneDAL algorithm has its own dedicated namespace, where corresponding
descriptor is defined. Descriptor, in its turn, defines a set of template
parameters and **properties** --- run-time parameters that can be accessed by
means of the corresponding getter and setter methods. `The following code sample
<descriptor-template_>`_ shows the common structure of descriptor's definition
for an abstract algorithm. To define a particular algorithm the following

> strings shall be substituted:

Reviewer: Not sure what to add as a definition of template parameters.

   class descriptor {
   public:
      /* Getter & Setter for the property called `%PROPERTY_NAME%` */
      descriptor& set_%PROPERTY_NAME%(%PROPERTY_TYPE% value);
      %PROPERTY_TYPE% get_%PROPERTY_NAME%() const;

      /* more properties */
   };

   } // namespace onedal::%ALGORITHM%

Each algorithm's meta-parameter is mapped to the property that shall satisfy the
following requirements:

- Properties are defined by means of getter and setter method. The underlying
  class field that stores the property's value is never exposed in the
  descriptor interface.

- The getter returns value of the underlying class field.

- The setter accepts the only parameter of property's type and assigns it
  to the underlying class field.

- Most of the properties are preset with default values, but some of them
  are initialized by passing required parameters to the constructor.

- The setter returns a reference to the descriptor object to allow chaining
  calls as shown in the example below.

  .. code-block:: cpp

     auto desc = descriptor{}
        .set_property_name_1(value_1)
        .set_property_name_2(value_2)
        .set_property_name_3(value_3);

> Descriptor is allowed to have any number of template parameters, but shall

Reviewer: ```suggestion
A descriptor is allowed to have any number of template parameters, but should
```

support at least two:

- ``Float`` is a `floating-point type <floating-point_>`_ that the algorithm
  uses for computations. This parameter is defined first and has the
  ``onedal::default_float_t`` default value.

> - ``Method`` is a tag-type that specifies an `computational method <methods_>`_.

Reviewer: ```suggestion
- ``Method`` is a tag-type that specifies the `computational method <methods_>`_.
```

  This parameter is defined second and has the ``method::by_default`` default

>   value.

Reviewer: I think those lines are logically belong to the part where template parameters are first introduced

.. _floating-point:

Floating-point Types
--------------------

> Algorithms are required to support at least one implementation-defined
> floating-point type. The other floating-point types, e.g., ``float``,
> ``double``, ``float16`` and ``bfloat16``, are optional and up to specific oneDAL
> implementation.

Reviewer: ```suggestion
It is required for each algorithm to support at least one implementation-defined
floating-point type. Other floating-point types are optional, for example ``float``,
``double``, ``float16``, and ``bfloat16``. It is up to a specific oneDAL
implementation whether or not to support these types.
```

The floating-point type used as a default in descriptors is

> implementation-defined and shall be declared within the top-level namespace.

Reviewer: ```suggestion
implementation-defined and should be declared within the top-level namespace.
```

.. code-block:: cpp

   namespace onedal {
      using default_float_t = /* implementation defined */;
   } // namespace onedal

.. _methods:

Computational Methods
---------------------

The supported computational methods are declared within the

> ``%ALGORITHM%::method`` namespace using tag-types. Algorithm shall support at

Reviewer: ```suggestion
``%ALGORITHM%::method`` namespace using tag-types. Each algorithm should support at
```

> least one computation method and declare the ``by_default`` type alias that
> refers to the one of the computational methods as shown in the example below.

Reviewer: ```suggestion
least one computational method and declare the ``by_default`` type alias that
refers to one of the computational methods as shown in the example below.
```

precision requirements vary depending on a workload.

To best tackle the mentioned challenges, each algorithm is decomposed into
`descriptors`_ and `operations`_.

.. _descriptors:

-----------
Descriptors
-----------

**A descriptor** is an object that represents an algorithm including all its
meta-parameters, dependencies on other algorithms, floating-point types, and
computational methods. A descriptor serves as:

- A dispatching mechanism for `operations`_. Based on a descriptor
  type, an operation executes a particular algorithm implementation.

- An aggregator of meta-parameters. It provides an interface for setting up
  meta-parameters at either compile-time or run-time.

- An object that stores the state of the algorithm. In the general case, a descriptor is
  a stateful object whose state changes after an operation is applied.

Each oneDAL algorithm has its own dedicated namespace, where the corresponding
descriptor is defined. Descriptor, in its turn, defines a set of template
parameters and **properties** --- run-time parameters that can be accessed by
means of the corresponding getter and setter methods. `The following code sample

> <descriptor-template_>`_ shows the common structure of descriptor's definition

Reviewer: ```suggestion
<descriptor-template_>`_ shows the common structure of a descriptor's definition
```

> for an abstract algorithm. To define a particular algorithm the following

Reviewer: ```suggestion
for an abstract algorithm. To define a particular algorithm, the following
```

capability to specify what type is needed is important for the end user as their
precision requirements vary depending on a workload.

To best tackle the mentioned challenges, each algorithm is decomposed into
`descriptors`_ and `operations`_.

.. _descriptors:

-----------
Descriptors
-----------

**A descriptor** is an object that represents an algorithm including all its
meta-parameters, dependencies on other algorithms, floating-point types, and
computational methods. A descriptor serves as:

- A dispatching mechanism for `operations`_. Based on a descriptor
  type, an operation executes a particular algorithm implementation.

- An aggregator of meta-parameters. It provides an interface for setting up
  meta-parameters at either compile-time or run-time.

- An object that stores the state of the algorithm. In the general case, a descriptor is
  a stateful object whose state changes after an operation is applied.

Each oneDAL algorithm has its own dedicated namespace (for more details, see
:ref:`common_namespaces`), where the corresponding descriptor is defined.
Descriptor, in its turn, defines defines the following:

> - **Template parameter.** A descriptor is allowed to have any number of template

Reviewer: Why not parameters? Below you have properties in plural

  parameters, but shall support at least two:

   + ``Float`` is a `floating-point type <floating-point_>`_ that the algorithm
     uses for computations. This parameter is defined first and has the
     ``onedal::default_float_t`` default value.

   + ``Method`` is a tag-type that specifies the `computational method
     <methods_>`_. This parameter is defined second and has the
     ``method::by_default`` default value.

> - **Properties.** A property is run-time parameter that can be accessed by means

Reviewer: is a run-time

- An object that stores the state of the algorithm. In the general case, a descriptor is
  a stateful object whose state changes after an operation is applied.

Each oneDAL algorithm has its own dedicated namespace (for more details, see
:ref:`common_namespaces`), where the corresponding descriptor is defined.

> Descriptor, in its turn, defines defines the following:

Reviewer: extra `defines`
